File: insertion-sort/insertion-sort2.js
const generateArray = (size) => {
    let arr = []
    for (let i = 1; i <= size; i++) {
        const index = Math.floor(Math.random() * i)
        arr.splice(index, 0, i)
    }
    return arr
}

const insertionSort = (arr) => {
    let res = []
    res[0] = arr[0]
    for (let i = 1; i < arr.length; i++) {
        let j = 0
        while (j < i && res[j] < arr[i]) j++
        if (j == res.length) {
            res.push(arr[i])
        } else {
            res.splice(j, 0, arr[i])
        }
    }
    return res
}

const insertionSort2 = (arr) => {
    for (let i = 1; i < arr.length; i++) {
        const key = arr[i];
        let j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
    return arr
}

const main = () => {
    const size = 30_000

    const genStart = new Date().getTime()
    const arr = generateArray(size)
    const genEnd = new Date().getTime()

    const sortStart = new Date().getTime()
    const res = insertionSort(arr)
    const sortEnd = new Date().getTime()

    const sort2Start = new Date().getTime()
    const res2 = insertionSort2(arr)
    const sort2End = new Date().getTime()

    const expected = []
    for (let i = 1; i <= arr.length; i++) expected.push(i)

    console.log(`Time to generate ${genEnd - genStart} ms`)

    if (expected.toString() != res.toString()) {
        console.error("Result1: They are not equal")
    } else {
        console.log("Result1: array sorted is as expected")
        console.log(`Result1: Time to sort ${sortEnd - sortStart} ms`)
    }

    if (expected.toString() != res2.toString()) {
        console.error("Result2: They are not equal")
    } else {
        console.log("Result2: array is sorted as expected")
        console.log(`Result2: Time to sort2 ${sort2End - sort2Start} ms`)
    }
}

main()

File: insertion-sort/insertion-sort2.cpp
#include <iostream>
#include <vector>

using namespace std;

int get_index(int i) { return rand() % i; }

vector<int> generate_array(int size)
{
    vector<int> arr = {};
    for (auto i = 1; i <= size; i++) {
        int index = get_index(i);
        arr.insert(arr.begin() + index, i);
    }
    return arr;
}

vector<int> insertion_sort(vector<int> arr)
{
    vector<int> res = {};
    res.push_back(arr[0]);
    for (auto i = 1; i < arr.size(); i++) {
        auto j = 0;
        while (j < i && res[j] < arr[i]) { j++; }
        res.insert(res.begin() + j, arr[i]);
    }
    return res;
}

vector<int> get_expected_array(int size)
{
    vector<int> arr = {};
    for (auto i = 1; i <= size; i++) {
        arr.push_back(i);
    }
    return arr;
}

bool compare_arrays(vector<int> subject, vector<int> expected)
{
    for (auto i = 0; i < subject.size(); i++) {
        if (subject[i] != expected[i]) return false;
    }
    return true;
}

double get_time_ms(clock_t start, clock_t end)
{
    auto time_sc = (double) (end - start) / CLOCKS_PER_SEC;
    return time_sc * 1000;
}

int main()
{
    const int size = 30000;

    // Generate array
    auto gen_start = clock();
    auto arr = generate_array(size);
    auto gen_end = clock();

    auto sort_start = clock();

    // Sort array
    auto res = insertion_sort(arr);

    // Eval sorted array with an expected result
    auto expected = get_expected_array(size);
    auto are_equal = compare_arrays(res, expected);

    auto sort_end = clock();

    // Output results
    cout << "1. Time to generate array: " << get_time_ms(gen_start, gen_end) << " ms" << endl;
    if (! are_equal) {
        cout << "2. Result array is NOT sorted as expected" << endl;
    } else {
        cout << "2. Result array is sorted as expected" << endl;
        cout << "3. Time to sort array: " << get_time_ms(sort_start, sort_end) << " ms" << endl;
    }

    return 0;
}
